ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -799,17 +963,33 @@ private void processEntryForLogSegmentBucket(ScanResults results, IndexEntry ind
     IndexValue indexValue = indexEntry.getValue();
     results.updateLogSegmentBaseBucket(indexValue.getOffset().getName(), indexValue.getSize());
     if (indexValue.isPut()) {
-      long expOrDelTimeInMs = indexValue.getExpiresAtMs();
+      long expiresAtMs = indexValue.getExpiresAtMs();
+      long deleteAtMs = Utils.Infinite_Time;
       IndexFinalState finalState = keyFinalStates.get(indexEntry.getKey());
       if (finalState != null && finalState.isDelete()) {
-        long deleteTimeInMs = finalState.getOperationTime();
-        expOrDelTimeInMs =
-            expOrDelTimeInMs != Utils.Infinite_Time && expOrDelTimeInMs < deleteTimeInMs ? expOrDelTimeInMs
-                : deleteTimeInMs;
+        deleteAtMs = finalState.getOperationTime();
+      }
+      if (expiresAtMs == Utils.Infinite_Time && deleteAtMs != Utils.Infinite_Time) {
+        handleLogSegmentDeletedBucketUpdate(results, indexValue, deleteAtMs, SUBTRACT);
+      } else if (expiresAtMs != Utils.Infinite_Time && deleteAtMs == Utils.Infinite_Time) {
+        handleLogSegmentExpiredBucketUpdate(results, indexValue, expiresAtMs, SUBTRACT);
+      } else if (expiresAtMs != Utils.Infinite_Time && deleteAtMs != Utils.Infinite_Time) {
+        if (expiresAtMs < deleteAtMs) {
+          handleLogSegmentExpiredBucketUpdate(results, indexValue, expiresAtMs, SUBTRACT);
+        } else {
+          handleLogSegmentDeletedBucketUpdate(results, indexValue, deleteAtMs, SUBTRACT);
+        }
       }
-      if (!indexValue.isTtlUpdate() || !isTtlUpdateEntryValid(indexEntry.getKey(), indexValue,
-          expOrDelTimeInMs == Utils.Infinite_Time ? expOrDelTimeInMs : expOrDelTimeInMs + 1, keyFinalStates)) {
-        handleLogSegmentBucketUpdate(results, indexValue, expOrDelTimeInMs, SUBTRACT);
+    } else if (!indexValue.isDelete() && !indexValue.isUndelete() && indexValue.isTtlUpdate()) {
```

Comment:
Since the `indexValue` is not from findKey() method, then I believe this can be simplified to `indexValue.isTtlUpdate()`.

Reply:
A DELETE or UNDELETE IndexValue can still return true when calling isTtlUpdate(), so there we have to do this kinda check.